### DISCUSSION TITLE
What is the right way of using driveFS?

### DISCUSSION THREAD

**Discussion by philipp-strack** [jupyterlite/jupyterlite#759] [Q&A]
Hello,

I obtained an emscripten file system instance FS from a binary and the corresponding PATH and ERRNO_CODES. I now want to mount driveFS in the emscripten filesystem to provide the binary with access to the jupyterlite files. What is the right way of achieving this?

I tried 

```
let baseUrl = PageConfig.getBaseUrl();
let mountpoint = '/drive';
let driveName = '';

const _drive = new DriveFS({ FS, PATH, ERRNO_CODES, baseUrl, driveName, mountpoint });
FS.mkdir(mountpoint);
FS.mount(_drive, '', mountpoint);
FS.chdir(mountpoint);
```

Is this the right approach? The above code works with no errors, but does not work. For example 

`FS.readdir(mountpoint);`

fails with

```
code: "EINVAL"
errno: 28
message: "Invalid argument"
node: undefined
```

Thanks a lot for your help!

**Comment by martinRenou** [comment]
Hello :)

This seems like the right approach.

Can I ask what your use-case is exactly? Maybe that will make it easier for helping.

I am not sure why it would not work in your case, but I may have some questions:
- what is the value of `driveName` in your case? Do you get this value from the kernel?
- did you register the service worker properly?

**Reply by philipp-strack** [reply]
Hello Martin,

Thanks a lot for your super quick reply.

> Can I ask what your use-case is exactly? Maybe that will make it easier for helping.

I want to run a wasm binary which was compiled by Emscripten and give it access to the files of jupyterlite. Such that the binary program which was written in c can make changes to the files present in jupyterlite.

> what is the value of driveName in your case? Do you get this value from the kernel?

 I set `drivename = ''` (I adjusted the question to reflect this). This is probably a mistake? How do I obtain the value from the kernel? Please excuse the naive question.

> did you register the service worker properly?

The answer is probably no. The above code is in its own typescript file and run as a worker with

`new Worker(new URL('./worker.js', import.meta.url), { type: 'module' });`

Is this the right way of running it?

Thanks again for your super quick reply.

**Reply by martinRenou** [reply]
So you are creating a custom kernel which makes use of the DriveFS?

You can probably look for guidance at https://github.com/jupyterlite/xeus-python-kernel which is a custom kernel which makes use of the DriveFS too.

> I set drivename = '' (I adjusted the question to reflect this). This is probably a mistake?

This is probably just fine for now. This will be the default value anyway for most JupyterLite users.

The drive name is the name of the JupyterLab drive, its value is different from `""` when you're using an extension like https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access (custom JupyterLab drive).

In pyolite this is inferred from the "location" of the kernel (the location of the Notebook starting the kernel):
- The kernel passes its location to the worker https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/src/kernel.ts#L76
- the worker infers the `driveName` from the location (first part of `split(':')`) https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/src/worker.ts#L149

> The answer is probably no. The above code is in its own typescript file and run as a worker with
`new Worker(new URL('./worker.js', import.meta.url), { type: 'module' });`
Is this the right way of running it?

I am not sure. You actually probably don't need to register the Worker yourself.

**Reply by philipp-strack** [reply]
Dear Martin, again thanks a lot for your quick reply. 

> So you are creating a custom kernel which makes use of the DriveFS?

Yes, we want to create a server extension which makes use of DriveFS

> This is probably just fine for now. This will be the default value anyway for most JupyterLite users.
> 
> The drive name is the name of the JupyterLab drive, its value is different from "" when you're using an extension like https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access (custom JupyterLab drive).
> 
> In pyolite this is inferred from the "location" of the kernel (the location of the Notebook starting the kernel):
> 
>     The kernel passes its location to the worker https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/src/kernel.ts#L76
>     the worker infers the driveName from the location (first part of split(':')) https://github.com/jupyterlite/jupyterlite/blob/main/packages/pyolite-kernel/src/worker.ts#L149

This explanation is very helpful thank you!

I will try to understand what the problem is and post here if I manage to resolve it.
